File: src/lib/webhooks.ts
import { invoke } from "@tauri-apps/api/core";

/**
 * Bridge for the outbound webhook egress feature (ContextForge #14).
 * Mirrors `~/.cortex/webhooks.json` via the Tauri commands defined in
 * `commands/webhooks.rs`.
 */

export interface Webhook {
  id: string;
  label: string;
  url: string;
  events: string[];
  headers: Record<string, string>;
  enabled: boolean;
}

export interface WebhookInput {
  id?: string;
  label: string;
  url: string;
  events: string[];
  headers: Record<string, string>;
  enabled: boolean;
}

export interface TestResult {
  ok: boolean;
  status: number | null;
  latency_ms: number;
  error: string | null;
}

/**
 * Curated list of events the agent runtime emits today. Users can still type
 * any custom event — this just powers autocomplete in WebhooksPanel.
 */
export const KNOWN_EVENTS: string[] = [
  "memory.snapshot.created",
  "memory.snapshot.restored",
  "task.complete",
  "task.failed",
  "approval.requested",
  "approval.granted",
  "session.start",
  "session.end",
  "agent.error",
];

export async function listWebhooks(): Promise<Webhook[]> {
  return invoke<Webhook[]>("list_webhooks");
}

export async function addWebhook(webhook: WebhookInput): Promise<Webhook> {
  return invoke<Webhook>("add_webhook", { webhook });
}

export async function updateWebhook(webhook: Webhook): Promise<void> {
  return invoke("update_webhook", { webhook });
}

export async function deleteWebhook(id: string): Promise<void> {
  return invoke("delete_webhook", { id });
}

export async function testWebhook(id: string): Promise<TestResult> {
  return invoke<TestResult>("test_webhook", { id });
}

export async function fireEvent(event: string, payload: unknown): Promise<number> {
  return invoke<number>("fire_event", { event, payload });
}

File: src-tauri/src/commands/webhooks.rs
//! Tauri command bridge for outbound webhooks. Thin wrappers over the
//! `observability::webhooks` module so the Rust side is unit-testable
//! without a tauri runtime.

use crate::observability::webhooks::{self, TestResult, Webhook, WebhookInput};

#[tauri::command]
pub async fn list_webhooks() -> Result<Vec<Webhook>, String> {
    webhooks::load_all()
}

#[tauri::command]
pub async fn add_webhook(webhook: WebhookInput) -> Result<Webhook, String> {
    webhooks::add(webhook)
}

#[tauri::command]
pub async fn update_webhook(webhook: Webhook) -> Result<(), String> {
    webhooks::update(webhook)
}

#[tauri::command]
pub async fn delete_webhook(id: String) -> Result<(), String> {
    webhooks::delete(&id)
}

#[tauri::command]
pub async fn test_webhook(id: String) -> Result<TestResult, String> {
    webhooks::test(&id)
}

#[tauri::command]
pub async fn fire_event(event: String, payload: serde_json::Value) -> Result<u32, String> {
    Ok(webhooks::fire(&event, &payload))
}
